Merging TOF reconstruction code to dev

chiarazampolli opened the pull request:
This is the PR to merge the code for the TOF reconstruction to dev. 
Tagging @noferini .

chiarazampolli committed:
Skeleton for matching

chiarazampolli committed:
Adding classes for TOF matching

chiarazampolli committed:
Further progress on TOF matching

chiarazampolli committed:
Further changes for TOF matching

chiarazampolli committed:
Taking the TPC track after the outwards propagation

Commit:
fix for compilation

chiarazampolli committed:
Propagate the TrackTPC corresponding to the track to match

chiarazampolli committed:
Getter and Setter of the value of the number of sigma to cut for the matching

chiarazampolli committed:
Use TPCITS track for TOF matching

chiarazampolli committed:
Fixes for matching (clusters that are loaded, filling of inxed cache...)

chiarazampolli committed:
Load geometry for DPL processing

chiarazampolli committed:
Load geometry if it is not there

chiarazampolli committed:
Temporary push during debugging for DPL

chiarazampolli committed:
Change readout window for DPL (and remove printout)

noferini committed:
TOF digitizer switch to vector of vectors for digits (#5)

chiarazampolli committed:
Progress on matching

Note that there is a lot of debug printout too in this commit.

To be still done:

- check how we write out the output
- time stored in clustered (we need the absolute one)
- further debug :)

noferini committed:
timestamp removed from TOF digits and other fixes (#6)

chiarazampolli committed:
Important changes in digitizer output and matching

- Digitizer output is now a vector of vectors (each "sub-vector" is per TOF readout frame)
- Matching now seems to work (but we still don't check the time)

chiarazampolli committed:
More changes for TOF matching - including macros to run and check it

noferini committed:
Tof rec (#8)

* move tof digitizer to vec of vecs also for MC labels

* ad tof test generator

chiarazampolli committed:
Removing macro with wrong name

chiarazampolli committed:
Further steps in TOF matching.

In this commit, there are many debug printing.
At this point, we reconstruct muons generated in a box, but we fail in PbPb.

chiarazampolli committed:
Fixing a few bugs.

The current version of teh code has a lot of debug printouts, which need to be
polish.
I am also committing some changes to the macros.

chiarazampolli committed:
One more bug fix, and a lot more.

We save also the labels now.
This commit contains a lot of debug printouts, and debug trees, to be removed.

chiarazampolli committed:
Polishing TOF matching code.

A lot of debug printing is possible.
Two macros added for debugging.

sawenzel commented on DataFormats/Detectors/TOF/include/DataFormatsTOF/Cluster.h:
```diff
@@ -40,23 +44,26 @@ class Cluster : public o2::BaseCluster<float>
-  float getPhi() const { return mPhi; } // Cluster Phi
+  float getR() const { return TMath::Sqrt(getX()*getX() + getY()*getY() + getZ()*getZ()); }     // Cluster Radius
+  float getPhi() const { return TMath::ATan2(getY(), getX()); } // Cluster Phi
+  int getSector() const { return int((TMath::ATan2(-getY(), -getX())+TMath::Pi())*TMath::RadToDeg()*0.05);} // Cluster Sector
```
is this function used often? If yes, it could potentially make sense to make the sector a property of the cluster instead of having to do this expensive operation over and over again.

chiarazampolli replied:
Changed, now we calculate sector, phi, r only once.

chiarazampolli committed:
Applying style

chiarazampolli committed:
Some replies to comments to the PR implemented

chiarazampolli commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -0,0 +1,801 @@
+  return false;
+}
+//______________________________________________
+void MatchTOF::loadTracksChunk(int chunk)
```
Removed. It was left from when we started implementing the class using the TPCITS matching as "template".

chiarazampolli commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -0,0 +1,801 @@
+  return false;
+}
+//______________________________________________
+void MatchTOF::loadTOFClustersChunk(int chunk)
```
Removed. It was left from when we started implementing the class using the TPCITS matching as "template".

chiarazampolli commented on DataFormats/Reconstruction/include/ReconstructionDataFormats/MatchInfoTOF.h:
```diff
@@ -0,0 +1,43 @@
+// or submit itself to any jurisdiction.
+
+/// \file Cluster.h
+/// \brief Definition of the TOF cluster
```
Yes, sorry. fixed now.

chiarazampolli commented on DataFormats/Reconstruction/src/MatchInfoTOF.cxx:
```diff
@@ -0,0 +1,18 @@
+// or submit itself to any jurisdiction.
+
+/// \file Cluster.cxx
+/// \brief Implementation of the TOF cluster
```
same :)

chiarazampolli committed:
Fix for compilation on mac

chiarazampolli commented:
> There is still a real compile error on MacOS

I know, but they don't come all in once, but after I fixed the first 2, this new one appeared. It is an incremental process, I cannot do different but fixing one at a time, and waiting the next check.

chiarazampolli committed:
Another fix for mac compilation - of course they are incremental fixes...

chiarazampolli commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -0,0 +1,801 @@
+    LOG(DEBUG) << "Global coordinates Before propagating to 371 cm: globalPos[0] = " << globalPos[0] << ", globalPos[1] = " << globalPos[1] << ", globalPos[2] = " << globalPos[2];
+    LOG(DEBUG) << "Radius xy Before propagating to 371 cm = " << TMath::Sqrt(globalPos[0]*globalPos[0] + globalPos[1]*globalPos[1]);
+    LOG(DEBUG) << "Radius xyz Before propagating to 371 cm = " << TMath::Sqrt(globalPos[0]*globalPos[0] + globalPos[1]*globalPos[1] + globalPos[2]*globalPos[2]);  
+    if (!propagateToRefX(trc, mXRef, 2) || TMath::Abs(trc.getZ()) > Geo::MAXHZTOF) {
```
There is already a margin.

chiarazampolli committed:
Making the tan of the width of a sector a constant

chiarazampolli commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -0,0 +1,801 @@
+      refReached = true; // we reached the 371cm reference
+    }
+    istep++;
+    if (fabs(trc.getY()) > trc.getX() * tan(o2::constants::math::SectorSpanRad / 2)) { // we are still in the same sector
```
Done, now it is a const at the beginning of the function.

chiarazampolli committed:
Following suggestion during PR - don't calculate sector, phi, r every time

Adding also a fix in a method name.

chiarazampolli committed:
fixes for build/o2checkcode/o2

chiarazampolli committed:
One more fix for the o2checkcode

chiarazampolli committed:
Updating debug printout

chiarazampolli committed:
git-clang-format change

sawenzel commented on DataFormats/Detectors/TOF/include/DataFormatsTOF/Cluster.h:
```diff
@@ -40,23 +42,41 @@ class Cluster : public o2::BaseCluster<float>
+  {
+    if (mR > 9990)
+      mR = TMath::Sqrt(getX() * getX() + getY() * getY() + getZ() * getZ());
+    return mR;
```
I have 2 more small comments here:
a) would it be possible to avoid hard coded constants (or at least explain the value of 9990)?
b) coding conventions state that every if statement should have the body enclosed within `{ }` to avoid strange bugs.

chiarazampolli replied:
Done.

chiarazampolli commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -0,0 +1,801 @@
+
+  mNumOfTracks = mTracksArrayInp->size();
+  if (mNumOfTracks == 0) return false; // no tracks to be matched
+  if (mMatchedTracksIndex) delete[] mMatchedTracksIndex;
```
No special reason. We guess it is more efficient than "new", so we followed your suggestion. Done.

chiarazampolli commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -0,0 +1,801 @@
+	  tofLabelSourceID[ilabel] = labelsTOF[ilabel].getSourceID();
+	}
+	//auto labelTPC = mTPCLabels->at(mTracksSectIndexCache[indices[0]][itrk]);    
+	auto labelITS = mITSLabels->at(mTracksSectIndexCache[indices[0]][itrk]);    
```
Agreed, we changed the 4 occurrences of "at".

chiarazampolli commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -0,0 +1,801 @@
+    mMatchedTracks.push_back(matchingPair); // array of MatchInfoTOF
+    const auto& labelTPC = mTPCLabels->at(matchingPair.first);
+    LOG(DEBUG) << "labelTPC: trackID = " << labelTPC.getTrackID() << ", eventID = " << labelTPC.getEventID() << ", sourceID = " << labelTPC.getSourceID();
+    const auto& labelITS = mITSLabels->at(matchingPair.first);
```
Done, see before.

noferini commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -0,0 +1,801 @@
+      }
+    }
+    int detIdTemp[5] = {-1, -1, -1, -1, -1}; // TOF detector id at the current propagation point
+    while (propagateToRefX(trefTrk, mXRef+istep*step, step) && nStripsCrossedInPropagation <=2 && mXRef+istep*step < Geo::RMAX){
```
we are going to use TrackPar for fast propagation.
I did't understand why TrackParCov::rotate method is not present in the mother class TrackPar, while I found TrackPar::rotatePar method. Since TrackParCov inherits from TrackPar I exepected the same name and the use of virtualization.

noferini replied:
I see. Thanks

chiarazampolli replied:
Hello Ruben,

We committed the modification. Now we use TrackPar until the inner radius of TOF. You suggested to use this approach till we mathched a strip, we can also do that if you think it is important. For now, please, have a look at how we propagate. 
Note that:
- we consider only the z component of the B field during the first "course" propagation;
- **as before**, the step of the propagation till 371 cm is 2 cm, only inside TOF it is 0.1 cm.

shahor02 replied:
Thanks @chiarazampolli , I added one more comment, but given that this small-step propagation may / pad_search  may be the most expensive operation, we better merge it is it is, and optimize it later, when we the outer param + length integration + propagator with mat.budget LUT is available.

chiarazampolli replied:
Ok, I resolve here now, we have the other comment open, in case you change your mind and prefer this changes now.

chiarazampolli committed:
Addressing comments in PR1486

In particular we changed the propagation: now we go to 371 cm using only
a TrackPar object, and using only the z component of the B field; if this
succeeds, we propagate again considering the errors.

chiarazampolli committed:
Leftover fixed

chiarazampolli committed:
git-clang-format changes

chiarazampolli committed:
git-clang-format - again

shahor02 commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -0,0 +1,828 @@
+    }
+    int detIdTemp[5] = { -1, -1, -1, -1, -1 }; // TOF detector id at the current propagation point
+    while (propagateToRefX(trefTrk, mXRef + istep * step, step) && nStripsCrossedInPropagation <= 2 && mXRef + istep * step < Geo::RMAX) {
+      if (0 && istep % 100 == 0) {
```
@chiarazampolli : Here also, I would not propagate from the full track in a 1mm steps. Instead, I would to smt. like:
```
TrackPar parOnly(trefTrk);
while (propagateToRefXNoCovariance( parOnly, mXRef + istep * step, step) ..., full field) {
    // check for strip crossing
    if ( .. crossed ) { // do full error 
         // do a few cm-size step full propagation to already found R.
    } 
}
}

Given that this part will certainly change once we start with outer track param, we can optimize it later.

chiarazampolli replied:
Hi @shahor02 , 

Yes, this is what I discussed with Francesco. Before changing it, we wanted to check with you if the way we did it in the first part was ok. There is no problem in changing here too, just let us know (in case you have a second thought :) ).

I have a question: suppose that we find a strip at 4.3 cm after 371 (just to say a number): when we say: "propagate to 371+4.3 in steps of 3 cm", will it do 1 step on 3 cm and 1 step of 1.3?

Chiara

sawenzel commented:
@chiarazampolli : Do you deem this ready to be merged?

chiarazampolli commented:
Ciao Sandro,

I think that we can work on the suggestions by Ruben in a separate PR, and close this one (he also said it is fine to do optimization later). 

As we discussed, we can squash all commits in one. 

Cheers,

Chiara
